ffcast.1.pod
=head1 NAME

ffcast - run command on rectangular screen regions

=head1 SYNOPSIS

=over 4

=item B<ffcast> [<options>] [<sub-command> [<args>]] [<command> [<args>]]

=back

=head1 DESCRIPTION

ffcast is a thin glue between interactive screen region selection and command
line tools that accept numerical geometry parameters. It makes passing
geometry parameters to such tools as easy as a mouse click.

Typically, ffcast prompts the user to select a screen region and injects the
geometry parameters thereof into the arguments of I<E<lt>commandE<gt>>, and
then executes the resultant command line. This injection is done via format
string substitution. See B<L<FORMAT STRINGS|/FORMAT STRINGS>>.

Multiple screen regions can be selected, and are by default combined by
"union" into a single region. If no valid region is specified by the user,
the root window is selected.

ffcast maintains an internal list of predefined B<sub-commands>, which is
looked up before all external commands. It's possible to add or override
sub-commands via user configuration. See B<L<SUB-COMMANDS|/SUB-COMMANDS>>.

If neither I<E<lt>sub-commandE<gt>> nor I<E<lt>commandE<gt>> is given, the
geometry of the selection is printed.

=head1 OPTIONS

Any of the I<E<lt>optionsE<gt>> can be repeated. They are parsed and applied
in the order they are specified.

=over 4

=item B<-#> I<E<lt>nE<gt>>

Select window with ID I<E<lt>nE<gt>>.

=item B<-b>

Include borders of windows to be selected.

=item B<-f>

Include window manager frame of windows to be selected. The window manager
must support B<_NET_FRAME_EXTENTS>.

=item B<-g> I<E<lt>geospecE<gt>>

Specify a selection in I<E<lt>geospecE<gt>>.
See B<L<GEOMETRY SPECIFICATION|/GEOMETRY SPECIFICATION>>.

=item B<-h>

Print help message, then exit.

=item B<-i>

Combine selected regions by intersection instead of union.

=item B<-q>

Be less verbose.

=item B<-s>

Select a rectangular region by dragging the mouse.

=item B<-v>

Be more verbose. B<-v> turns on verbose messages. B<-vv> turns on debug.

=item B<-w>

Select a window by clicking the mouse in that window.

=item B<-x> I<E<lt>nE<gt>>

Select a B<Xinerama> head of ID I<E<lt>nE<gt>>. Multiple heads can be
specified either as a comma-separated list, or by repeating this option. Use
C<list> or C<l> to get a list of head IDs and the corresponding geometry
parameters.

=back

=head1 FORMAT STRINGS

All I<E<lt>argsE<gt>> to an external I<E<lt>commandE<gt>> are always subject
to format string substitution.

Note I<E<lt>argsE<gt>> to an internal I<E<lt>sub-commandE<gt>> are not
automatically subject to format string substitution, although sub-commands may
choose to do format string substitution on their own.

The following format strings are supported:

=over 4

=item B<%D>

B<DISPLAY>

=item B<%h>

Height

=item B<%w>

Width

=item B<%x>

Left offset

=item B<%y>

Top offset

=item B<%X>

Right offset

=item B<%Y>

Bottom offset

=item B<%c>

Equivalent to B<%x,%y>

=item B<%C>

Equivalent to B<%X,%Y>

=item B<%g>

Equivalent to B<%wx%h+%x+%y>

=item B<%s>

Equivalent to B<%wx%h>

=item B<%%>

A literal %

=back

=head1 GEOMETRY SPECIFICATION

A I<E<lt>geospecE<gt>> specifies a geometry in one of the following forms:

=over 4

=item S<I<x>B<,>I<y> I<X>B<,>I<Y>>

I<x>, I<y>, I<X> and I<Y> are the left, top, right and bottom offsets,
respectively.

=item I<w>B<x>I<h>B<+>I<x>B<+>I<y>

I<w>, I<h>, I<x> and I<y> are the width, height, left offset and top offset,
respectively.

=back

=head1 SUB-COMMANDS

The first normal argument is looked up first in the list of sub-commands. If
there is no match, it is executed as an external command.

To bypass sub-command lookup entirely, use B<%> I<command>.

To get a list of all sub-commands, use B<help> without an argument. B<help>
I<sub-command> prints help for the specified sub-command.

To import sub-commands, ffcast B<source>s the files I<@LIBDIR@/ffcast/subcmd>,
I</etc/ffcast/subcmd> and I<$XDG_CONFIG_HOME/ffcast/subcmd>, in that order.

To learn how to register a sub-command, refer to I<@LIBDIR@/ffcast/subcmd>.

=head1 EXAMPLES

List all sub-commands with brief description:

    ffcast help

Use the sub-command B<png> to take a screenshot of a window, including its
window manager frame. Note the B<png> sub-command supports format string
substitution:

    ffcast -fw png screenshot-%s.png

The B<-f> option works only if the window manager supports
B<_NET_FRAME_EXTENTS>. A workaround is to use the B<pad> sub-command to
explicitly pad the region with the size of the frame:

    ffcast -w pad '20 2 2' png screenshot_%s.png

Use the sub-command B<rec> to record a fullscreen screencast on display C<:1>.
The B<rec> sub-command accepts the B<-m> option, which crops the video so
both its width and height are divisible by the given number:

    DISPLAY=:1 ffcast -vv rec -m 4 screencast.mkv

Record a screencast of all three windows of B<GIMP>. This works because all
selections are by default combined by union. Note format strings are always
substituted for external commands:

    ffcast -www ffmpeg -f x11grab -show_region 1 -s %s -i %D+%c gimp.mp4

Resize a window to the same size as another, by two mouse clicks:

    ffcast -w xdotool selectwindow windowsize %w %h windowfocus

Take a screenshot of two windows combined, excluding anything invisible on
head 0, using either of:

    ffcast -i -x 0 -g "$(ffcast -ww)" png
    ffcast -ww ffcast -i -x 0 -g %g png

Take three individual screenshots of heads 0, 1, 2, one by one. The B<each>
command runs a command for each user selection consecutively:

    ffcast -x 0,1,2 each png head_%g.png

=head1 AUTHOR

lolilolicon E<lt>lolilolicon@gmail.comE<gt>

=head1 SEE ALSO

B<ffmpeg(1)>, B<X(7)>

=for comment
vim:ft=pod:spell:spelllang=en_us:cc=78:

=cut


src/ffcast.bash
#!/bin/bash
#
# ffcast @VERSION@
# Copyright (C) 2011-2014  lolilolicon <lolilolicon@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

if (( BASH_VERSINFO[0] == 4 && BASH_VERSINFO[1] < 3 )) ||
   (( BASH_VERSINFO[0] < 4 )); then
    printf 'fatal: requires bash 4.3+ but this is bash %s\n' "$BASH_VERSION"
    exit 43
fi >&2

set -e +m -o pipefail
shopt -s extglob lastpipe
trap -- 'trap_err $LINENO' ERR

readonly -a srcdirs=(
    '@pkglibexecdir@'
    '@sysconfdir@/@PACKAGE@'
    "${XDG_CONFIG_HOME:-$HOME/.config}"/'@PACKAGE@')
readonly -a logl=(error warn msg verbose debug)
declare -A 'logp=([warn]="warning" [msg]=":")'
declare -- verbosity=2
declare -A sub_commands=() sub_cmdfuncs=()
declare -a rects=() regions=()
declare -A heads=() windows=() heads_all=()
declare -- {root_{w,h},rect_{w,h,x,y,X,Y}}=0
declare -- borders=0 frame=0 frame_extents_support=1 intersection=0

declare -A fmtmap=(
    ['D']='$DISPLAY'
    ['h']='$rect_h'
    ['w']='$rect_w'
    ['x']='$rect_x'
    ['y']='$rect_y'
    ['X']='$rect_X'
    ['Y']='$rect_Y'
    ['c']='$rect_x,$rect_y'
    ['C']='$rect_X,$rect_Y'
    ['g']='${rect_w}x$rect_h+$rect_x+$rect_y'
    ['s']='${rect_w}x$rect_h')

#---
# Functions

msg_colors_on() {
    logp[error]=$'\e[1;31m''error'$'\e[m'
    logp[warn]=$'\e[1;33m''warning'$'\e[m'
    logp[msg]=$'\e[34m'':'$'\e[m'
    logp[verbose]=$'\e[32m''verbose'$'\e[m'
    logp[debug]=$'\e[36m''debug'$'\e[m'
}

trap_err() {
    set -- "$1" "${PIPESTATUS[@]}"
    printf '%s:%d: ERR:' "${BASH_SOURCE[0]}" "$1"; shift
    printf ' PIPESTATUS:'
    printf ' %d' "$@"
    printf '  BASH_COMMAND: %s\n' "$BASH_COMMAND"
} >&2

_msg() {
    local prefix=$1
    shift || return 0
    local fmt=$1
    shift || return 0
    printf '%s' "$prefix"
    printf -- "$fmt\n" "$@"
}

_quote_cmd_line() {
    local prefix=$1
    shift || return 0
    local cmd=$1
    shift || return 0
    printf '%s' "$prefix"
    printf '%q' "$cmd"
    (( $# )) && printf ' %q' "$@"
    printf '\n'
}

_report_array_by_key() {
    local -n ref_array=$1
    printf '%q[%q]=%q\n' "$1" "$2" "${ref_array[$2]}"
}

debug_array_by_key() {
    (( verbosity >= 4 )) || return 0
    printf '%s: ' "${logp[debug]}"
    _report_array_by_key "$@"
} >&2

for ((i=0; i<${#logl[@]}; ++i)); do
    eval "${logl[i]}() {
        (( verbosity >= $i )) || return 0
        _msg \"\${logp[${logl[i]}]-${logl[i]}}: \" \"\$@\"
    } >&2"
done

for ((i=3; i<${#logl[@]}; ++i)); do
    eval "${logl[i]}_dryrun() {
        (( verbosity >= $i )) || return 0
        _quote_cmd_line \"\${logp[${logl[i]}]-${logl[i]}}: cmdline: \" \"\$@\"
    } >&2
    ${logl[i]}_run() {
        ${logl[i]}_dryrun \"\$@\" && \"\$@\"
    }"
done

# $1: array variable of format string mappings
# $2: array variable to assign substitution results to
# ${@:3} are strings to be substituted
substitute_format_strings() {
    local -n ref_fmtmap=$1
    local -n ref_strarr=$2
    shift 2
    local noref_str
    ref_strarr=()
    while (( $# )); do
        noref_str=
        printf '%s' "$1" |
        while IFS= read -r -n 1 -d ''; do
            if [[ $REPLY == '%' ]]; then
                IFS= read -r -n 1 -d '' || :
                if [[ -v ref_fmtmap[$REPLY] ]]; then
                    eval "noref_str+=${ref_fmtmap[$REPLY]}"
                elif [[ $REPLY == '%' ]]; then
                    noref_str+='%'
                else
                    noref_str+="%$REPLY"
                fi
            else
                noref_str+=$REPLY
            fi
        done
        ref_strarr+=("$noref_str")
        shift
    done
}

printf '%s %s\n' max '>' min '<' | while IFS=' ' read -r mom cmp; do
    eval 'get_'$mom'_offsets() {
        local offsets=$1 o
        shift || return 1
        local {,_}{l,t,r,b}
        IFS=" " read l t r b <<< "$offsets"
        for offsets in "$@"; do
            [[ -n $offsets ]] || continue
            IFS=" " read _{l,t,r,b} <<< "$offsets"
            for o in l t r b; do
                eval "(( (_$o '$cmp' $o) && ($o = _$o) )) || :"
            done
        done
        printf "%d %d %d %d\n" "$l" "$t" "$r" "$b"
    }'
done
unset -v mom cmp

# $1: a geospec
# $2: variable to assign offsets to
set_region_by_geospec() {
    local noref_region=$2
    local IFS
    # sanitize whitespaces
    IFS=$' \t'; set -- $1; set -- "$*"
    case $1 in
        ?(-)+([0-9])+(\ |,)?(-)+([0-9])+(\ |,)?(-)+([0-9])+(\ |,)?(-)+([0-9]))  # x1,y1 x2,y2
            IFS=' ,'
            set -- $1
            ;;
        +([0-9])x+([0-9])\+?(-)+([0-9])\+?(-)+([0-9]))  # wxh+x+y
            IFS='x+'
            set -- $1
            set -- "$3" "$4" "$((root_w - $3 - $1))" "$((root_h - $4 - $2))"
            ;;
        *)
            return 1
            ;;
    esac
    IFS=' '
    printf -v "$noref_region" '%s' "$*"
}

# $1: variable to assign offsets to
set_region_interactively() {
    msg "%s" "please select a region using mouse"
    printf -v "$1" '%s' "$(xrectsel '%x %y %X %Y')"
}

# $1: array variable to modify
# $2: variable to assign window ID to
set_window_interactively() {
    msg "%s" "please click once in target window"
    (( !frame || frame_extents_support )) || set -- "$1" "$2" -frame
    xwininfo_get_window_by_ref "$@"
}

# $1: a window ID
# $2: array variable to modify
# $3: variable to assign window ID to
set_window_by_id() {
    # Unlike xprop, xwininfo simply ignores an invalid -id argument
    if [[ $1 != @(+([0-9])|0x+([[:xdigit:]])) ]] ||
        [[ $(printf '%d' "$1") == 0 ]]; then
        error 'invalid window id: %s' "$1"
        return 1
    fi
    xwininfo_get_window_by_ref "$2" "$3" -id "$1"
}

# stdin: xdpyinfo -ext XINERAMA (preferably sanitized)
# $1: array variable to assign heads to, i.e. =([id]=offsets ...)
xdpyinfo_get_heads_by_ref() {
    local -n ref_heads=$1
    local IFS
    while IFS=' ' read -r REPLY; do
        REPLY=${REPLY#head #}
        if [[ $REPLY == +([0-9]):\ +([0-9])x+([0-9])' @ '+([0-9]),+([0-9]) ]]; then
            IFS=' :x@,'
            set -- $REPLY
            set -- "$1" "$4" "$5" "$((root_w - $4 - $2))" "$((root_h - $5 - $3))"
            IFS=' '
            ref_heads["$1"]="${*:2}"
        fi
    done
    (( $# == 5 ))
}

xdpyinfo_list_heads() {
    LC_ALL=C xdpyinfo -ext XINERAMA | grep '^  head #' | sed 's/^ *//'
}

# stdin: xwininfo output (locale: C)
# stdout: wxh
xwininfo_get_size() {
    local IFS
    set --
    while IFS=$' \t' read -r REPLY; do
        if [[ $REPLY == 'Width: '+([0-9]) ]]; then
            set -- "${REPLY#'Width: '}" "$@"
        elif [[ $REPLY == 'Height: '+([0-9]) ]]; then
            set -- "$@" "${REPLY#'Height: '}"
        else
            continue
        fi
        if (( $# == 2 )); then
            IFS=x
            printf '%s\n' "$*"
            return
        fi
    done
    return 1
}

# $1: array variable to modify
# $2: variable to assign window ID to
# ${@:3} are passed to xwininfo
xwininfo_get_window_by_ref() {
    local -n ref_windows=$1
    local -n ref_id=$2
    export LC_ALL=C
    xwininfo "${@:3}" |
    awk -v borders="$borders" -v frame="$((frame && frame_extents_support))" '
    BEGIN { OFS = " " }
    /^xwininfo: Window id: 0x[[:xdigit:]]+ / { _id = $4 }
    /^ *Border width: [[:digit:]]+$/ { _bw = $3 }
    $1 == "Corners:" && NF == 5 && split($2, a, /\+/) == 3 {
        _ol = a[2]
        _ot = a[3]
        if (split($3, a, /\+/) == 2) {
            _or = substr(a[1], 2)
            _ob = substr($4, length(a[1]) + 2)
        }
    }
    END {
        if (_id == "" || _bw == "" || _ob == "")
            exit 1
        if (frame) {
            xprop = "xprop -id \"" _id "\" -notype _NET_FRAME_EXTENTS"
            while ((xprop | getline) && ($1 != "_NET_FRAME_EXTENTS"));
            close(xprop)
            if ($1 == "_NET_FRAME_EXTENTS") {
                sub(/.*= /, "")
                split($0, a, /[ ,]+/)
                _ol -= a[1]; _ot -= a[3]; _or -= a[2]; _ob -= a[4]
            }
        }
        else if (!borders) {
            _ol += _bw; _ot += _bw; _or += _bw; _ob += _bw
        }
        print _id
        print _ol, _ot, _or, _ob
    }' |
    {
        read -r
        ref_id=$REPLY
        read -r
        ref_windows["$ref_id"]=$REPLY
    }
}

run_default_command() {
    printf '%dx%d+%d+%d\n' "$rect_w" "$rect_h" "$rect_x" "$rect_y"
}

run_external_command() {
    local -- cmd=$1 extcmd
    shift || return 0
    local -a args
    # always substitute format strings for external commands
    substitute_format_strings fmtmap args "$@"
    # make sure it's an external command -- a disk file
    if ! extcmd=$(type -P "$cmd"); then
        error "external command '%s' not found" "$cmd"
        return 127
    fi
    verbose_run command -- "$extcmd" "${args[@]}"
}

run_subcmd_or_command() {
    local sub_cmd=$1
    if [[ -z $sub_cmd ]]; then
        run_default_command
        exit
    fi
    if [[ -v sub_commands[$sub_cmd] ]]; then
        shift
        local sub_cmd_func=${sub_cmdfuncs[$sub_cmd]:-$sub_cmd}
        if [[ $(type -t "$sub_cmd_func") == function ]]; then
            verbose_run "$sub_cmd_func" "$@"
        else
            error "sub-command '%s' function '%s' not found" "$sub_cmd" \
                "$sub_cmd_func"
            exit 1
        fi
    else
        run_external_command "$@" || exit
    fi
}

set_region_vars_by_offsets() {
    offsets=$(get_max_offsets "$offsets" '0 0 0 0')
    debug "sanitize offsets -> offsets='%s'" "$offsets"
    IFS=' ' read rect_{x,y,X,Y} <<< "$offsets"
    (( rect_w = root_w - rect_x - rect_X )) || :
    (( rect_h = root_h - rect_y - rect_Y )) || :
    set -- rect_{w,h,x,y,X,Y}
    debug 'set region variables'
    while (( $# )); do
        debug '\t%s' "$(declare -p "$1")"
        shift
    done
    if ! (( rect_w > 0 && rect_h > 0 )); then
        error 'invalid region size: %sx%s' "$rect_w" "$rect_h"
        return 1
    fi
}

#---
# Process command line options and rectangles

[[ ! -t 2 ]] || msg_colors_on

usage() {
    cat <<EOF
ffcast @VERSION@
Usage:
  ${0##*/} [options] [sub-command [args]] [command [args]]

Options:
  -g <geospec>  specify a region in numeric geometry
  -x <n|list>   select the Xinerama head of ID n
  -s            select a rectangular region by mouse
  -w            select a window by mouse click
  -# <n>        select a window by window ID
  -b            include window borders hereafter
  -f            include window frame hereafter
  -i            combine regions by intersection
  -q            be less verbose
  -v            be more verbose
  -h            print this help and exit

All options can be repeated, and are processed in order.
If no region is selected by the user, select fullscreen.

For more details see ffcast(1).
EOF
  exit "${1:-0}"
}

LC_ALL=C xwininfo -root | xwininfo_get_size | IFS=x read root_{w,h} || usage 1

declare -- i=0 id= opt= var=
declare -a ids
OPTIND=1
while getopts ':#:bfg:hiqsvwx:' opt; do
    case $opt in
        h)  usage;;
        x)
            [[ $OPTARG != l?(ist) ]] || { xdpyinfo_list_heads; exit; }
            # cache list of all heads once
            if (( ! ${#heads_all[@]} )); then
                if ! xdpyinfo_list_heads |
                    xdpyinfo_get_heads_by_ref heads_all; then
                    error 'failed to get all Xinerama heads'
                    exit 1
                fi
                debug 'got all Xinerama heads'
                debug '\t%s' "$(declare -p heads_all)"
            fi
            if [[ $OPTARG == all ]]; then
                ids=("${!heads_all[@]}")
            else
                IFS=' ,' read -a ids <<< "$OPTARG"
            fi
            for id in "${ids[@]}"; do
                if [[ $id != +([0-9]) ]]; then
                    warn "ignored invalid head ID: \'%s'" "$id"
                elif [[ ! -v heads_all[$id] ]]; then
                    warn "ignored non-existent head ID: \`%s'" "$id"
                else
                    heads[$id]=${heads_all[$id]}
                    rects[i++]="heads[$id]"
                fi
            done
            ;;
        g)
            var="regions[${#regions[@]}]"
            if ! set_region_by_geospec "$OPTARG" "$var"; then
                warn "ignored invalid geospec: \`%s'" "$OPTARG"
            else
                rects[i++]=$var
            fi
            ;;
        s)
            var="regions[${#regions[@]}]"
            set_region_interactively "$var"
            rects[i++]=$var
            ;;
       \#)
            set_window_by_id "$OPTARG" windows id || exit
            var="windows[$id]"
            rects[i++]=$var
            ;;
        w)
            set_window_interactively windows id
            var="windows[$id]"
            rects[i++]=$var
            ;;
        b)
            borders=1
            verbose "windows: now including borders"
            ;;
        f)
            frame=1
            verbose "windows: now including window manager frame"
            if ! LC_ALL=C xprop -root -notype _NET_SUPPORTED |
                grep -qw _NET_FRAME_EXTENTS; then
                frame_extents_support=0
                warn 'no _NET_FRAME_EXTENTS support; using xwininfo -frame'
            fi
            ;;
        i)  intersection=1;;
        q)  (( (verbosity > 0) && verbosity-- )) || :;;
        v)  (( (verbosity < ${#logl[@]} - 1) && verbosity++ )) || :;;
      '?')  warn "invalid option: \`%s'" "$OPTARG";;
      ':')  error "option requires an argument: \`%s'" "$OPTARG"; exit 1;;
    esac
done
shift $(( OPTIND - 1 ))

#---
# Combine all rectangles

declare -- mom offsets=
declare -n ref_rect

(( intersection )) && mom=max || mom=min

for ref_rect in "${rects[@]}"; do
    offsets=$(get_"$mom"_offsets "$ref_rect" "${offsets}")
    debug "get_%s_offsets -> offsets='%s'" "$mom" "$offsets"
done
: ${offsets:='0 0 0 0'}
unset -n ref_rect
unset -v mom

set_region_vars_by_offsets || exit

#---
# Import predefined sub-commands

# a little optimization
(( $# )) || { run_default_command; exit; }

for srcdir in "${srcdirs[@]}"; do
    subcmdsrc=$srcdir/subcmd
    if [[ -r $subcmdsrc ]]; then
        verbose "importing sub-commands from file %s" "$subcmdsrc"
        . "$subcmdsrc"
    fi
done
unset -v srcdir subcmdsrc

# make sure these are not defined as sub-commands
for cmd in builtin command; do
    unset -f $cmd
    if [[ -v sub_commands[$cmd] ]]; then
        unset -v sub_commands[$cmd]
        warn 'unset sub-command %s' "$cmd"
    fi
done
unset -v cmd

#---
# Execute

run_subcmd_or_command "$@"

# vim:ts=4:sw=4:et:cc=80:
